### ISSUE TITLE
kube-proxy/iptables, UDP conntrack data not removed on endpoint changes

### ISSUE THREAD

**Issue by joemiller** [kubernetes/kubernetes#59368]
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
When deleting/recreating pods backing a UDP service, traffic from external senders stops flowing.

**What you expected to happen**:
Traffic would continue to flow to new pods.

**How to reproduce it (as minimally and precisely as possible)**:
Based on reading of similar issues and commit messages it seems like this would be most noticable on high-volume UDP applications, statsd and syslog are good examples.

1. Create a deployment and service for a UDP app. In my test case, this would be the `atlassianlabs/gostatsd:5.2.0` container.
2. On an external node, create a small test script utilizing a statsd library such as pystatsd. The most important part is that the sender uses the same source-port for every packet sent. Ex:
```
import time
import pystatsd

stats = pystatsd.Client('104.197.127.53', 8125)
while True:
    stats.increment('joe-testing-delete-me-timing.testing')
    time.sleep(0.2)
```
3. Run the sender. Install tcpdump into the pod and watch for incoming packets, example from the test case: `tcpdump -A -s0 port 8125 | grep joe-testing`
4. Delete the pod
5. Run tcpdump in the newly created pod. Observe that no new matching packets are flowing.
6. Clear existing conntrack rules on all nodes in the cluster. You should see traffic start flowing to the pod again. example command for clearing conntrack rules on the nodes: `docker run --net=host --privileged --rm claesjonsson/conntrack -D -p udp --orig-port-dst 8125`

**Anything else we need to know?**:
I have tested this only on GKE-managed clusters that utilize google forwarding-rules for external LBs, but I believe the issue may be present on any platform where iptables mode for kube-proxy is used. Tested and verified on 1.8.6-gke.0 and 1.7.11-gke.1. Also tested with both Google Container-optimized OS and the Ubuntu-based node images.

I have only test this on GKE 

It appears this was a known issue and should have been cleared up by: https://github.com/kubernetes/kubernetes/pull/22573/files 

From my reading of iptables/proxier.go, it seems as if the `deleteEndpointConnections()` func should exec() `conntrack` to clear up UDP entries on service endpoint change, but the only place I see that func called is from syncProxyRules(). That func looks complex and possibly like it is only run occasionally and not on service/endpoint changes.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.6", GitCommit:"6260bb08c46c31eea6cb538b34a9ceb3e406689c", GitTreeState:"clean", BuildDate:"2017-12-21T06:34:11Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.6-gke.0", GitCommit:"ee9a97661f14ee0b1ca31d6edd30480c89347c79", GitTreeState:"clean", BuildDate:"2018-01-05T03:36:42Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}
```
(Also tested on 1.7.11-gke.1)
- Cloud provider or hardware configuration:
GKE-managed clusters, google cloud.
- OS (e.g. from /etc/os-release):
tried both:
```
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
and
```
BUILD_ID=10032.71.0
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=c4c6234ae4f384ce00819c41b48ca8f6f1fa3ba8
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=63
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=63
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos
```
- Kernel (e.g. `uname -a`):
```
Linux gke-cluster-01-kube17-04b-6782975f-4xwn 4.4.86+ #1 SMP Thu Dec 7 20:11:11 PST 2017 x86_64 Intel(R) Xeon(R) CPU @ 2.60GHz GenuineIntel GNU/Linux
```

and

```
Linux gke-udp-ubuntu-test-default-pool-cf65ad93-bpnb 4.4.0-1027-gke #27-Ubuntu SMP Thu Aug 10 13:13:08 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
- Others:

**Comment by jeffspahr** [comment]
We also ran into this in kubernetes v1.7.6 using OpenShift 3.7:

https://bugzilla.redhat.com/show_bug.cgi?id=1537780

**Comment by pdecat** [comment]
I believe this is the same as https://github.com/kubernetes/kubernetes/issues/59033

If so, it is fixed in 1.10: https://github.com/kubernetes/kubernetes/pull/59286

**Comment by shparker1977** [comment]
Is there any other known workaround for this besides updating conntrack rules on the kube nodes?  I tried deleting and recreating the service, but this didn't update the rules.  When I restart the stream on the client side data flows correctly again (perhaps it's perceived as a new connection), but I have to restart the stream for each sender.

**Comment by squeed** [comment]
FWIW, this is not fixed in 1.10 - that fix was for HostPort, and this is still a bug for NodePort.

**Comment by mhristache** [comment]
I am seeing this issue in 1.11. I am using LoadBalancer type.

**Comment by squeed** [comment]
@JacobTanenbaum has been working on a fix for this, FWIW.

**Comment by samos123** [comment]
This issue is extremely impactful for telephony protocols that rely on static source ip and source port. Any update on a possible workaround?

I was thinking a potential workaround for this specific use case of UDP traffic it might be fine to disable conntrack on the K8s nodes [1]. UDP is a connectionless protocol anyway. Curious to hear your thoughts whether this is stupid. Another idea might be to reduce the UDP timeout to something really short so the stale conntrack entries go away faster.

[1] Disabling conntrack for udp: https://jeanbruenn.info/2017/04/30/conntrack-and-udp-dns-with-iptables/

**Comment by squeed** [comment]
The workaround is to not use NodePorts. Other kinds of services are already correct.

**Comment by mhristache** [comment]
@squeed not sure what you mean. I am seeing the issue with a LoadBalancer type so it does not affect NodePorts only.

@samos123 As a workaround I have created a small python script that watches the K8s API for pod deletions and clears the related conntrack entries when that happens. I can share it when I get to work on Monday if you want to try it out.

**Comment by mhristache** [comment]
I've added my script [here](https://github.com/mhristache/k8s-conntrack-cleanup). 

Note that it's not tested much so don't use it in a production environment. Also it cleans up all conntrack entries for a deleted pod, not only the UDP related ones.

**Comment by tderoock** [comment]
We are also seeing this issue with type loadbalancer. We are on 1.11.

Our use-case is the following:
multiple devices are sending to an aggreagator machine. This machine forwards the UDP packages to a loadBalancer in gcp. 

When debugging the service, I noticed the following:
- The traffic is assigned to one pod, and the packages are not loadbalanced (client and server are the same cfr use-case)
- When redeploying a new cluster, Conntrack is cleared on termination signal of the old pod. (client receives connection refused)
- red/black and highlander deployment strategies have the same effect.

How would I expect this to be working then?
- New cluster is added.
- Conntrack is cleared
- Data is routed to NEW pods
- old pods are terminated

Best Regards,

Tim

**Comment by abhinbalur** [comment]
I am seeing this  exact same issue on our cluster which has k8s v1.14.5 . 
We have endpoint nginx pods backing a UDP service with nodeport.
Suddenly one day we see traffic not hitting into nginx pod but reaching the node hosting the nginx pod.
We could simulate the problem by restarting the nginx pods.
I am assuming with this issue marked as closed; and having made its way into v1.14;  we should not be seeing this.
Would also like to point out; the one thing, in addition we have in the service is ExternalTrafficPolicy set to Local.
Can anyone please guide/help on why we are seeing this in v.1.14.5 of our cluster.

**Comment by hulucc** [comment]
Seeing this with k8s v1.16.5
